File: libraries/react-native/src/core/utils/account-token.ts
import { sha1 } from "@noble/hashes/legacy.js";
import { utf8ToBytes } from "@noble/hashes/utils.js";
import { Effect } from "effect";

/** Shared UUIDv5 namespace for StoreKit and Google Play account identifiers. */
export const VOIDHASH_ACCOUNT_TOKEN_NAMESPACE = "3919eb4e-3466-593c-8c1e-84554e13a0a6";

const uuidToBytes = (uuid: string): Uint8Array => {
  const hex = uuid.replaceAll("-", "");
  if (!/^[0-9a-fA-F]{32}$/.test(hex)) {
    return Effect.runSync(Effect.die(new TypeError(`Invalid UUID namespace: ${uuid}`)));
  }

  const bytes = new Uint8Array(16);
  for (let index = 0; index < bytes.length; index += 1) {
    bytes[index] = Number.parseInt(hex.slice(index * 2, index * 2 + 2), 16);
  }
  return bytes;
};

const bytesToUuid = (bytes: Uint8Array): string => {
  const hex = Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0")).join("");
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
};

/** Derives an RFC 4122 UUIDv5 from a namespace UUID and UTF-8 name. */
export const uuidV5 = (namespaceUuid: string, name: string): string => {
  const namespaceBytes = uuidToBytes(namespaceUuid);
  const nameBytes = utf8ToBytes(name);
  const input = new Uint8Array(namespaceBytes.length + nameBytes.length);
  input.set(namespaceBytes);
  input.set(nameBytes, namespaceBytes.length);

  const bytes = sha1(input).slice(0, 16);
  bytes[6] = ((bytes[6] as number) & 0x0f) | 0x50;
  bytes[8] = ((bytes[8] as number) & 0x3f) | 0x80;
  return bytesToUuid(bytes);
};

/** Derives the deterministic account token supplied to StoreKit and Google Play. */
export const deriveAccountToken = (distinctId: string): string =>
  uuidV5(VOIDHASH_ACCOUNT_TOKEN_NAMESPACE, distinctId);

File: libraries/ios/Sources/VoidhashCore/Identity/AccountToken.swift
import CryptoKit
import Foundation

/// Deterministic account identifiers supplied to StoreKit and Google Play.
///
/// The derivation is a cross-platform contract: the backend resolves purchases by recomputing
/// the same UUIDv5, so the namespace, hash and lowercase formatting must match
/// `src/core/utils/account-token.ts` byte for byte.
public enum AccountToken {
    /// Shared UUIDv5 namespace for StoreKit and Google Play account identifiers.
    public static let namespace = "3919eb4e-3466-593c-8c1e-84554e13a0a6"

    /// Derives an RFC 4122 UUIDv5 from a namespace UUID and a UTF-8 name.
    ///
    /// - Returns: The lowercase UUID string, or `nil` when `namespaceUuid` is not a UUID.
    public static func uuidV5(namespaceUuid: String, name: String) -> String? {
        guard let namespaceBytes = uuidToBytes(namespaceUuid) else {
            return nil
        }

        var input = namespaceBytes
        input.append(contentsOf: Array(name.utf8))

        var bytes = Array(Insecure.SHA1.hash(data: Data(input)).prefix(16))
        bytes[6] = (bytes[6] & 0x0f) | 0x50
        bytes[8] = (bytes[8] & 0x3f) | 0x80
        return bytesToUuid(bytes)
    }

    /// Derives the deterministic account token for a distinct id.
    public static func derive(distinctId: String) -> String {
        // The namespace is a compile-time constant UUID, so the parse can never fail.
        return uuidV5(namespaceUuid: namespace, name: distinctId) ?? ""
    }

    private static func uuidToBytes(_ uuid: String) -> [UInt8]? {
        let hex = uuid.replacingOccurrences(of: "-", with: "")
        guard hex.count == 32 else {
            return nil
        }

        var bytes: [UInt8] = []
        bytes.reserveCapacity(16)
        var index = hex.startIndex
        while index < hex.endIndex {
            let next = hex.index(index, offsetBy: 2)
            guard let byte = UInt8(hex[index..<next], radix: 16) else {
                return nil
            }
            bytes.append(byte)
            index = next
        }
        return bytes
    }

    private static func bytesToUuid(_ bytes: [UInt8]) -> String {
        let hex = bytes.map { String(format: "%02x", $0) }.joined()
        let slice = { (start: Int, end: Int) -> String in
            let lower = hex.index(hex.startIndex, offsetBy: start)
            let upper = hex.index(hex.startIndex, offsetBy: end)
            return String(hex[lower..<upper])
        }
        return [slice(0, 8), slice(8, 12), slice(12, 16), slice(16, 20), slice(20, 32)]
            .joined(separator: "-")
    }
}
